Add FAQ on ACL change of ownership (#2062)
rcgsheffield/sheffield_hpc: FAQs.rst

   **Total**                                            **52**
   ==============================================       =========================== 

How to change ownership of files/folders not as root?
-----------------------------------------------------

> Normal users will not have the permission to **chown** since they are not being root or in sudoers. Therefore, the only valid method for transferring ownership of files/folders is to make

Reviewer: We try to be succinct where possible. Maybe something like:
Cluster users can not use **chown** for transferring ownership of files/folders, however it is possible using Access Contol Lists (ACLs).   
The continue with: The following assumes...

the documents readable to the person who is taking ownership, reset the ownership, copy, delete and finally move the documents.

This process will also be useful for users who end up with an older account and a newer account. It allows user to transfer data across local filesystems from older account to the newer.
Please note that the below guide assumes that both accounts are still be active. If you have lost access to the old account in the last few weeks then get in touch with us via research-it@sheffield.ac.uk and we may be able to help transfer files across.

> The following assumes user1 is the current owner and user2 is going to be the new owner. Please change the ownership of the files/folders using the following instructions:

Reviewer: ```suggestion
The following assumes user1 is the current owner and user2 is going to be the new owner:
```

> 1. user1 makes the files/folders to be changed ownership available to read by user2, most likely with Linux ACLs:

Reviewer: ```suggestion
1. user1 makes the files/folders available to read by user2 with Linux ACLs:
```

.. code-block:: bash

        setfacl --recursive --modify u:user2:r-x /the/directory/changing/ownership

2. user2 copies file from there:

.. code-block:: bash

        cp -R /the/directory/changing/ownership my/tmp/directory

3. user2 checks if they have the copy of the file with the correct ownership:

.. code-block:: bash

        ls -l

> 

Reviewer: give an example output

4. user1 deletes the existing folder recursively:

.. code-block:: bash

        rm -rf /the/directory/changing/ownership

> 5. user2 moves the folder back where it was before:

Reviewer: ```suggestion
5. user2 moves the folder back to it's original location:
```

   **Total**                                            **52**
   ==============================================       =========================== 

How to change ownership of files/folders not as root?
-----------------------------------------------------

> Cluster users cannot use **chown** for transferring ownership of files/folders. However, it is possible using Access Control Lists (ACLs).

Reviewer 2: For this part I think we will want to explain a little more background or users may have further unanswered questions. I was thinking about:

- The 'chown' command can only be successfully executed by the root user on any linux system(why? security - https://unix.stackexchange.com/questions/27350/why-cant-a-normal-user-chown-a-file/27374#27374)

- HPC users are not allowed to use the root account or assume its permissions (why? security on a shared system like HPC)

- Explain in human terms what we're doing before we give CLI type instructions e.g. we will bypass these limitations by giving the 2nd user read permissions on the data so they can make a copy of their own, then we can delete the original data.

   **Total**                                            **52**
   ==============================================       =========================== 

How to change the ownership of files and folders when not the root user?
------------------------------------------------------------------------

> Cluster users are not allowed to use the root account for system administrators to control and ensure the security of all users.

Reviewer: ```suggestion
Cluster users are not allowed to use the root account due to it's elevated privileges, it is reserved for system administrators to ensure the security of all users.
```

Reviewer 2: Here I think we probably want reword a bit to indicate 

> "Only system administrators are granted access to the root account (superuser privileges) for security reasons."

Can possibly link on `superuser privileges`  out to: https://en.wikipedia.org/wiki/Superuser

Side note on "it's" versus "its": the former contraction is equivalent to "it is", so should only be used where that works.

> Cluster users are not allowed to use the root account for system administrators to control and ensure the security of all users.

Reviewer 3: You are not wrong but I think "HPC users" would more clear for newbies. Ill let you make the call though.

> Thus, they cannot use **chown** for transferring ownership of files/folders due to the possibility of leading to security vulnerabilities.

Reviewer: ```suggestion
Thus, they cannot use **chown** for transferring ownership of files/folders.
```

Reviewer 2: Think we're also missing some background on: "As using chown requires owning the file or being root for security reasons, thus it is not possible for a non-root user to reassign ownership."

Where "security reasons" can link out to https://unix.stackexchange.com/a/27374

However, it is possible using Access Control Lists (ACLs).

> The files/folders have to be stored in public Fastdata areas, detailed instructions are contained in the :ref:`fastdata_dir`.

Reviewer: Could have this as item 1 on the list. I.e make sure user2 has access to the folder/files

1. **user1** checks the original permissions of the files/folders:

.. code-block:: console

        [user1@login1 [stanage] public]$ getfacl the/directory/changing/ownership/
        # file: the/directory/changing/ownership/
        # owner: user1
        # group: clusterusers
        user::rwx
        group::r-x
        other::r-x

2. **user1** makes the files/folders available to read by **user2** with Linux ACLs:

.. code-block:: console

        setfacl --recursive --modify u:user2:r-x the/directory/changing/ownership/

3. **user1** ensures **user2** has the access to the files/folders:

.. code-block:: console
        :emphasize-lines: 3, 9

        [user1@login1 [stanage] public]$ ls -l
        total 4
        drwxrwxr-x+ 2 user1 clusterusers 4096 Apr  3 13:52 the/directory/changing/ownership/
        [user1@login1 [stanage] public]$ getfacl  the/directory/changing/ownership/
        # file: the/directory/changing/ownership/
        # owner: user1
        # group: clusterusers
        user::rwx
        user:user2:r-x
        group::r-x
        mask::r-x
        other::r-x

4. **user2** copies the files from **user1**:

.. code-block:: console

>         cp -R /mnt/parscratch/users/user1/public/the/directory/changing/ownership/ my/tmp/directory

Reviewer 4: If someone is following this through they probably won't have a tmp dir. Maybe add a line 'mkdir my/tmp/directory/'. I'd just make it part of step 2, no need for an extra step.

Most users will probably know to make a directory but I think it will make for a more streamlined process for users following it through. :)

   **Total**                                            **52**
   ==============================================       =========================== 

How to change the ownership of files and folders when not the root user?
------------------------------------------------------------------------

> Only system administrators are granted access to the root account (`superuser privileges <https://en.wikipedia.org/wiki/Superuser>`_) for security reasons.
> As using the **chown** command requires owning the file or being root for `security reasons <https://unix.stackexchange.com/a/27374>`_, it is not possible for a non-root user to reassign ownership.
> However, it is achievable using Access Control Lists (ACLs).

Reviewer 2: ```suggestion
For security reasons only system administrators are granted access to the root account (`superuser privileges <https://en.wikipedia.org/wiki/Superuser>`_)
and as successfully using the **chown** command `requires root account permissions <https://unix.stackexchange.com/a/27374>`_ it is not possible for a non-root user to directly reassign ownership.
However it is possible to do so indirectly by using Access Control Lists (ACLs).
```

4. **user1** ensures **user2** has the access to the files/folders:

.. code-block:: console
        :emphasize-lines: 3, 9

        [user1@login1 [stanage] public]$ ls -l
        total 4
        drwxrwxr-x+ 2 user1 clusterusers 4096 Apr  3 13:52 the/directory/changing/ownership/
        [user1@login1 [stanage] public]$ getfacl  the/directory/changing/ownership/
        # file: the/directory/changing/ownership/
        # owner: user1
        # group: clusterusers
        user::rwx
        user:user2:r-x
        group::r-x
        mask::r-x
        other::r-x

5. **user2** creates a temporary directory to store the files:

.. code-block:: console

        mkdir my/tmp/directory

6. **user2** copies the files from **user1**:

.. code-block:: console

        cp -R /mnt/parscratch/users/user1/public/the/directory/changing/ownership/ my/tmp/directory

7. **user2** checks if they have the copy of the files/folders with the correct ownership:

.. code-block:: console

        [user2@login1 [stanage] public]$ ls -l
        total 4
        drwxr-xr-x 2 user2 clusterusers 4096 Apr  3 14:00 the/directory/changing/ownership

> 8. **user1** deletes the existing folder recursively:

Reviewer 2: We might want a warning admonishment here that they should check with user2 the files have transferred prior to deletion as the fastdata areas do not have backups.
